ddl/ddl_tiflash_api.go

```diff
@@ -346,37 +366,148 @@ func updateTiFlashStores(pollTiFlashContext *TiFlashManagementContext) error {
+	pollMaxCount := PollAvailableTableProgressCount
+	failpoint.Inject("PollAvailableTableProgressMaxCount", func(val failpoint.Value) {
+		pollMaxCount = uint64(val.(int))
+	})
+	for element := pollTiFlashContext.AvailableTables.Front(); element != nil && pollMaxCount > 0; pollMaxCount-- {
+		availableTableID := element.Value.(AvailableTableID)
+		var table table.Table
+		if availableTableID.IsPartition {
+			table, _, _ = schemas.FindTableByPartitionID(availableTableID.ID)
+			if table == nil {
+				logutil.BgLogger().Info("get table by partition failed, may be dropped or truncated",
+					zap.Int64("partitionID", availableTableID.ID),
+				)
+				pollTiFlashContext.AvailableTables.Remove(element)
+				element = element.Next()
+				continue
+			}
+		} else {
+			var ok bool
+			table, ok = schemas.TableByID(availableTableID.ID)
+			if !ok {
+				logutil.BgLogger().Info("get table id failed, may be dropped or truncated",
+					zap.Int64("tableID", availableTableID.ID),
+				)
+				pollTiFlashContext.AvailableTables.Remove(element)
+				element = element.Next()
+				continue
+			}
+		}
+
+		tableInfo := table.Meta()
+		if tableInfo.TiFlashReplica == nil {
+			logutil.BgLogger().Info("table has no TiFlash replica",
+				zap.Int64("tableID or partitionID", availableTableID.ID),
+				zap.Bool("IsPartition", availableTableID.IsPartition),
+			)
+			pollTiFlashContext.AvailableTables.Remove(element)
+			element = element.Next()
+			continue
+		}
+		progress, err := getTiFlashTableSyncProgress(pollTiFlashContext, availableTableID.ID, tableInfo.TiFlashReplica.Count)
+		if err != nil {
+			logutil.BgLogger().Error("get tiflash sync progress failed",
+				zap.Error(err),
+				zap.Int64("tableID", availableTableID.ID),
+				zap.Bool("IsPartition", availableTableID.IsPartition),
+			)
+			continue
+		}
+		if pollTiFlashContext.ProgressCache[availableTableID.ID] != progress {
+			err = infosync.UpdateTiFlashTableSyncProgress(context.Background(), availableTableID.ID, progress)
+			if err != nil {
+				logutil.BgLogger().Error("updating TiFlash replica process failed",
+					zap.Error(err),
+					zap.Int64("tableID or partitionID", availableTableID.ID),
+					zap.Bool("IsPartition", availableTableID.IsPartition),
+					zap.String("progress", progress),
+				)
+				continue
+			}
+			pollTiFlashContext.ProgressCache[availableTableID.ID] = progress
+		}
+		pollTiFlashContext.AvailableTables.Remove(element)
+		element = element.Next()
+	}
+}
+
+func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContext *TiFlashManagementContext) error {
 	defer func() {
-		pollTiFlashContext.HandlePdCounter++
-		pollTiFlashContext.HandlePdCounter %= PullTiFlashPdTick.Load()
+		failpoint.Inject("PollTiFlashReplicaStatusCleanProgressCache", func() {
+			pollTiFlashContext.PollCounter = PollCleanProgressCacheInteal
+		})
+		pollTiFlashContext.PollCounter++
```

Comment:
How about sharing the same ticker with UpdateTiFlashStoreCounter?

The ticker pattern usually structured as below, which is more clear and more simplified to add / remove tick actions:

```
fn tickFn() {
    tick++;
    if (tick % actionAInterval) == 0 {
        doActionA();
    }
    if (tick % actionBInterval) == 0 {
        doActionB();
    }
    if (tick % actionCInterval) == 0 {
        doActionC();
    }
}

loop {
    tickFn();
    sleep(tickInterval);
}
```

Reply:
BTW I do worry about the impact of a single action takes too long time, when they are now sharing the same ticker.